_pytest/config.py

```diff
@@ -1257,9 +1257,13 @@ def get_common_ancestor(args):
 
 
 def get_dirs_from_args(args):
-    return [d for d in (py.path.local(x) for x in args
-                        if not str(x).startswith("-"))
-            if d.exists()]
+    possible_dirs = [
+        py.path.local(str(arg).split('::')[0])
+        for arg in args
+        if not str(arg).startswith('-')
+    ]
+
+    return [d for d in possible_dirs if d.exists()]
```

Comment:
I played around with a couple of variants of this. I was not a big fan of the original nested generator expression in the list comprehension and it became worse with this additional check.

The args may not be `py.path.local` from the beginning. Also, not stripping the part after `::` before calling `.exists()` leaves us with the original bug. To do everything in a single list comprehension it would then look like
```py
   return [
        py.path.local(str(arg).split('::')[0])
        for arg in args
        if not str(arg).startswith('-') and py.path.local(str(arg).split('::')[0]).exists()
    ]
```

I was not happy with that since stripping `::` and creating the `py.path.local` object happens twice.